src/vs/base/common/comparers.ts

```diff
@@ -39,48 +177,188 @@ export function noIntlCompareFileNames(one: string | null, other: string | null,
+	const [oneName, oneExtension] = extractNameAndExtension(one);
+	const [otherName, otherExtension] = extractNameAndExtension(other);
+	const collator = intlFileNameCollator.getValue().collator;
+	const collatorCaseInsensitive = intlFileNameCollatorCaseInsensitive.getValue().collator;
+
+	// Check for case insensitive differences in extensions.
+	let result = collatorCaseInsensitive.compare(oneExtension, otherExtension);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Check for case differences in names.
+	let caseResult = compareCaseLowerFirst(oneName, otherName);
+	if (caseResult !== 0) {
+		return caseResult;
+	}
+
+	// Check for other differences in names.
+	result = collator.compare(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Disambiguate extension case if needed.
+	if (oneExtension !== otherExtension) {
+		return collator.compare(oneExtension, otherExtension);
+	}
+
+	return 0;
+}
+
+/** Compares filenames by extension, then by name. */
+export function compareFileExtensionsMixed(one: string | null, other: string | null): number {
+	const [oneName, oneExtension] = extractNameAndExtension(one);
+	const [otherName, otherExtension] = extractNameAndExtension(other);
+	const collator = intlFileNameCollator.getValue().collator;
+	const collatorCaseInsensitive = intlFileNameCollatorCaseInsensitive.getValue().collator;
+
+	// Check for case insensitive differences in extensions.
+	let result = collatorCaseInsensitive.compare(oneExtension, otherExtension);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Check for differences in names.
+	result = collator.compare(oneName, otherName);
+	if (result !== 0) {
+		return result;
+	}
+
+	// Disambiguate extension case if needed.
+	if (oneExtension !== otherExtension) {
+		return collator.compare(oneExtension, otherExtension);
+	}
+
+	return 0;
+}
+
+/** Compares filenames by extension unicode value, then by name unicode value. */
+export function compareFileExtensionsUnicode(one: string | null, other: string | null) {
+	const [oneName, oneExtension] = extractNameAndExtension(one);
+	const [otherName, otherExtension] = extractNameAndExtension(other);
+
+	// Check for extension differences
+	if (oneExtension !== otherExtension) {
+		return oneExtension < otherExtension ? -1 : 1;
+	}
+
+	// Check for name differences.
+	if (oneName !== otherName) {
+		return oneName < otherName ? -1 : 1;
+	}
+
+	return 0;
+}
+
+const FileNameMatch = /^(.*?)(\.([^.]*))?$/;
+
+/** Extracts the name and extension from a full filename, with special handling for dotfiles */
+export function extractNameAndExtension(str?: string | null): [string, string] {
 	const match = str ? FileNameMatch.exec(str) as Array<string> : ([] as Array<string>);
 
-	return [(match && match[1]) || '', (match && match[3]) || ''];
+	let result: [string, string] = [(match && match[1]) || '', (match && match[3]) || ''];
+
+	// treat an empty filename with an extension, or a filename that starts with a dot, as a dotfile name
```

Comment:
Added this special handling of dotfiles as mentioned in the PR submission comment.